903: Implement trace-only mode for vt

JacobDomagala opened the pull request:
Fixes #903 

### Build vt:

> export VT_BUILD_TRACE_ONLY=1
> /vt/ci/build_cpp.sh /vt /build

or

> cmake -Dvt_trace_only=1

### Example usage: 

I've used https://github.com/DARMA-tasking/vt-sample-project for testing
```cmake
cmake_minimum_required(VERSION 3.10 FATAL_ERROR)

project(vt-sample)

set(CMAKE_CXX_STANDARD 14)

find_package(vt REQUIRED)

file(
  GLOB
  HEADER_FILES
  ${CMAKE_CURRENT_SOURCE_DIR}/src/*.h
)

file(
  GLOB
  SOURCE_FILES
  ${CMAKE_CURRENT_SOURCE_DIR}/src/*.cc
)

add_executable(
  vt-sample
  ${HEADER_FILES} ${SOURCE_FILES}
)

target_link_libraries(vt-sample PUBLIC vt::vt-trace)
```


```c
#include <mpi.h>
#include "vt/trace/trace_lite.h"

void doWork(MPI_Comm comm) {
  // Find out rank, size
  int world_rank;
  MPI_Comm_rank(comm, &world_rank);
  int world_size;
  MPI_Comm_size(comm, &world_size);

  int number;
  if (world_rank == 0) {
    number = -1;
    MPI_Send(&number, 1, MPI_INT, 1, 0, comm);
  } else if (world_rank == 1) {
    MPI_Recv(&number, 1, MPI_INT, 0, 0, comm, MPI_STATUS_IGNORE);
  }
}

int main(int argc, char** argv) {
  MPI_Init(&argc, &argv);

  ::vt::trace::TraceLite myTrace("TestApp");

  myTrace.initializeStandalone(MPI_COMM_WORLD);

  doWork(MPI_COMM_WORLD);

  myTrace.flushTracesFile();
  myTrace.finalizeStandalone();

  MPI_Finalize();
}
```

Output for running above example on 2 ranks:

- node/rank 0

> PROJECTIONS-RECORD 0
> 6 0
> 100 12058624 205 1 0 0
> 100 12058624 206 1 0 0
> 100 172490752 2197 2 0 0
> 100 172490752 2540 2 0 0
> 7 228412

- node/rank 1

> PROJECTIONS-RECORD 0
> 6 0
> 100 8912896 198 1 1 0
> 100 8912896 204 1 1 0
> 100 172490752 2662 2 1 0
> 100 172490752 3099 2 1 0
> 7 3543

Where:
8912896     -> _MPI_Recv 
12058624   -> _MPI_Send 
172490752 -> trace_flush


It's also possible to use existing API for tracing such as :
- enableTracing
- disableTracing

nmm0 reviewed (approved):
Looks good to me! Quick question -- would it be straightforward to have a minimal header to include just the trace API? I think it would help reducing compile times not having to include runtime headers.

PhilMiller commented:
The code changes generally look fine to me.

The diff coverage check is a reminder that we don't seem to have suitably targeted testing for the tracing subsystem. Perhaps that should be added.

nmm0 commented on src/CMakeLists.txt:
```diff
@@ -109,70 +108,104 @@ list(APPEND PROJECT_SUBDIRS_LIST ${TOP_LEVEL_SUBDIRS})
+    get_filename_component( DIR ${FILE} DIRECTORY )
+    install(
+      FILES ${FILE}
+      DESTINATION "include/${DIR}"
```
A weird cmake-ism -- DIR here contains the absolute path so installing is installing the headers to install-dir/include/absolutepath/ so the path is broken.

nmm0 reviewed (changes requested):
Additionally, could you change it from replacing the vt library to adding a separate trace library (maybe enabled by the cmake flag)? This way is simpler for the end user. In our use case in NimbleSM it would make our build system logic significantly easier.

JacobDomagala commented:
@nmm0  I've updated the PR (also updated the description on top with example CMake usage). Let me know if there's anything you'd want to be added. 
In the meantime i will try to create CI tests for vt-trace target and will try to add some documentation.

lifflander reviewed (commented):
I'm seeing this error when trying to compile with trace-only:

```
[ 86%] Built target nimble
[ 90%] Building CXX object src/CMakeFiles/NimbleSM_MPI.dir/nimble_mpi_main.cc.o
In file included from /Users/jliffla/codes/collision/NimbleSM/src/nimble_mpi_main.cc:55:
In file included from /Users/jliffla/codes/vt/vt-install-clang-9/include/vt-trace/vt/trace/trace_lite.h:51:
In file included from /Users/jliffla/codes/vt/vt-install-clang-9/include/vt-trace/vt/trace/trace_user_event.h:56:
In file included from /Users/jliffla/codes/vt/vt-install-clang-9/include/vt-trace/vt/context/context.h:54:
/Users/jliffla/codes/vt/vt-install-clang-9/include/vt-trace/vt/utils/tls/tls.h:53:12: fatal error: 'vt/utils/tls/omp_tls.h' file not found
  #include "vt/utils/tls/omp_tls.h"
           ^~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

I think that @cz4rs's work on disabling threading might be useful here for the trace-only mode where we don't need threading at all.

lifflander commented:
Threading should probably just be disabled by default. Or, we could just include those threading files (but I think this is unnecessary).

lifflander commented:
I pushed some changes. Once @jstrzebonski gets the vt-sample-project working, let's add the trace-only build to it for testing.

jstrzebonski commented:
@JacobDomagala @lifflander 
#1230 is merged

JacobDomagala commented on ci/build_cpp.sh:
```diff
@@ -215,7 +216,7 @@ fi
 if test "$is_alpine" -eq 0 && test "${VT_CI_BUILD:-0}" -eq 1 && test "${target}" = "install"
 then
-    git clone https://github.com/DARMA-tasking/vt-sample-project
+    git clone https://github.com/DARMA-tasking/vt-sample-project -b sample-use-of-trace-only-vt
```
Temporarily using non-master branch

JacobDomagala replied:
Will be changed to master once https://github.com/DARMA-tasking/vt-sample-project/issues/1 is merged

JacobDomagala commented on src/vt/runtime/component/diagnostic_value.h:
```diff
@@ -388,15 +388,16 @@ struct DiagnosticSnapshotValues {
 };
 
-/**
+/**1
```
Fixed

nmm0 commented on src/CMakeLists.txt:
```diff
@@ -263,6 +253,12 @@ link_target_with_vt(
 # Define and set up the different build types of VT
+
+if (vt_trace_only)
+  include(../cmake/trace_only_functions.cmake)
```
I think this include line needs to be pulled outside the if statement. There are some functions later that are defined in this file and CMake errors out if vt_trace_only is set to OFF

JacobDomagala replied:
That's kinda weird. This file only defines two functions (create_trace_only_target and set_trace_only_config) and both are called only when vt_trace_only is set to 1.

JacobDomagala replied:
Anyways I've updated the PR.

nmm0 reviewed (changes requested):
Just a minor change

lifflander reviewed (approved):
Looks good

nmm0 reviewed (approved):
Works for me!
